File: src/main/java/dev/reeve/mazelib/MazeDirection.java
package dev.reeve.mazelib;

import lombok.Getter;
import lombok.val;

import java.util.ArrayList;
import java.util.List;

@Getter
public enum MazeDirection {
	EAST(1, 0),
	SOUTH(0, 1),
	WEST(-1, 0),
	NORTH(0, -1);
	
	private final int x;
	private final int y;
	
	MazeDirection(int x, int y) {
		this.x = x;
		this.y = y;
	}
	
	public static List<MazeDirection> getDirections(boolean east, boolean south, boolean west, boolean north) {
		val directions = new boolean[]{east, south, west, north};
		val list = new ArrayList<MazeDirection>();
		
		for (int i = 0; i < 4; i++) {
			if (directions[i])
				list.add(MazeDirection.values()[i]);
		}
		
		return list;
	}
}

File: src/main/kotlin/dev/reeve/mazelib/MazeDirection.kt
package dev.reeve.mazelib

import dev.reeve.mazelib.MazeDirection
import java.util.ArrayList

/**
 * The directions you can look at neighbors with using things like [MazePosition.inDirection]
 * Order is starting at EAST, and going clockwise (ordinal of these are very important to a few functions like [MazePoint.setOppositeOpen].
 */
enum class MazeDirection(val x: Int, val y: Int) {
	EAST(1, 0), SOUTH(0, 1), WEST(-1, 0), NORTH(0, -1);
	
	companion object {
		@JvmStatic
		fun getDirections(east: Boolean, south: Boolean, west: Boolean, north: Boolean): List<MazeDirection> {
			val directions = booleanArrayOf(east, south, west, north)
			val list = ArrayList<MazeDirection>()
			for (i in 0..3) {
				if (directions[i]) list.add(values()[i])
			}
			return list
		}
	}
}